docs: RFC-106 - Record Level and Secondary Index Support for Flink Writers (#17610)
apache/hudi: rfc/rfc-102/rfc-102.md

  the License.  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
# RFC-102: RLI support for Flink streaming

## Proposers

- @danny0405

## Approvers
 - @geserdugarov
 - @vinothchandar
 - @cshuo

## Status

GH Discussion: https://github.com/apache/hudi/discussions/17452

> Please keep the status updated in `rfc/README.md`.

## Abstract
This RFC aims to introduce RLI support for Flink streaming:

- Impl reliable and performant write and read support for RLI via Flink APIs;
- The RLI impl is engines compatible, for e.g, Flink can access and utilize the RLI written by Spark and vice versa;
- The RLI is global, upserts among partitions is supported; Also support partition level RLI for large fact tables;
- Async compaction for MDT when RLI is enabled; in writer pipeline or table services background job;
- Smart caching of RLI;

> - Clearly set limits for the kind of write throughput supported by RLI (based on certain average response time for the RLI access, like from x0ms to x00ms) via empirical benchmarks;

Reviewer: Do you mean some recommendations for users when to use RLI? But if you mean to set some limits in the code on our side, then I suppose that it should be done by Flink with backpressure mechanism.

> - Clearly set limits for the kind of write throughput supported by RLI (based on certain average response time for the RLI access, like from x0ms to x00ms) via empirical benchmarks;

Author: > I suppose that it should be done by Flink with backpressure mechanism.

backpressure can handle the regular volumn fluctuate instead of the continupus lagging for processing, which would cause stability issues though.

- Ability to be expanded to arbitrary secondary indexing on different columns.

## Background
Flink does not support RLI while spark does, this caused inconsistency between engines, for tables migrated from Spark to flink streaming, the index type needs to be switched to either bucket or flink_state , this caused a overhead for users in production.

> Another reason is for multiple partition upserts, currently the only choice is flink_state index, but the flink_state actually costs a lot of memory and can not be shared between different workloads.

Author: yeah

## Implementation

The high-level ideas:

- a RLI based index backend will be there to replace the flink_state index;
- a cache of RLI would be introduced to speed the access;
- a separate index function to write the RLI/SI payloads;
- the MDT RLI files is written synchronously with the data table data files, the metadata is sent to the coordinator for a final commit to the MDT(after `FILES` partition is ready);
- the MDT compaction is switched to be async and the data files compaction pipeline is reused for less take up of task slots.

### The Write

### The RLI Access

> In `BucketAssigner` operator, the RLI index metadata would be utilized as the index backend, the `BucketAssigner` operator will probe the RLI with the incoming record keys to figure out whether msg is update or insert or delete.

Reviewer 2: currently the partitioner for bucket assigner with flink state index is also simply hashing by key, which may also has skew issue. Anyway, we can benchmark it to get specific numbers.

> In `BucketAssigner` operator, the RLI index metadata would be utilized as the index backend, the `BucketAssigner` operator will probe the RLI with the incoming record keys to figure out whether msg is update or insert or delete.

Reviewer: Should we use `keyBy()` before `BucketAssigner` operator?

With cache layer there is possible wrong tagging:
- `checkpoint1` is completed
- Hudi committed T1
- `record5` is absent in RLI
- `record5` arrived and sent to `BucketAssigner0`, here we saved info about this record in cache
- `checkpoint2` is completed
- async start of Hudi commit T2
- `record5` arrived and sent to `BucketAssigner1`, where cache miss will happen, and there is no info in RLI due to Hudi commit T2 in progress.
`record5` will be tagged as insert, but should be update.

> In `BucketAssigner` operator, the RLI index metadata would be utilized as the index backend, the `BucketAssigner` operator will probe the RLI with the incoming record keys to figure out whether msg is update or insert or delete.

Reviewer 2: @geserdugarov yeah, it's a basic requirement to shuffle records with same key into the same sub-task of bucket assigner to ensure the correct tagging. 

What we're talking about here is, should we shuffle different keys under the same file group of RLI in metadata table into the same sub-task, since the hoodie keys will first be grouped by the RLI file group before querying the index.

> In `BucketAssigner` operator, the RLI index metadata would be utilized as the index backend, the `BucketAssigner` operator will probe the RLI with the incoming record keys to figure out whether msg is update or insert or delete.

Reviewer 3: > should we shuffle different keys under the same file group of RLI in metadata table into the same sub-task

We should be reading reach RLI shard once per commit . Even with caching, its a terrible idea to not group keys to task for RLI lookup, since it ll affect cache efficiency. i.e each node has to then cache all RLI shards.. which can be massive.

> In `BucketAssigner` operator, the RLI index metadata would be utilized as the index backend, the `BucketAssigner` operator will probe the RLI with the incoming record keys to figure out whether msg is update or insert or delete.

Author: +1

In other words, the RLI index metadata will serve as the same role of the `flink_state` index.

#### The Cache of RLI Access
We need fast access in streaming to have high throughput(ideally per record access should be < 10ms), thus a general hotspot cache is needed. We will build a in-memory LRU cache by the active upsert records keys, the cache items will be force evictted by a configured memory threshold.

> We also need a memory cache for the index mappings of current checkpoint because it is not committed to Hudi table yet so invisible.

Reviewer 2: the mappings should be persisted somewhere in case the failover happens and memory cache will be lost.

Author: not really, they are persisted in the MDT, that's why we read inglight commits from MDT.

Reviewer 4: so you meant a buffer of current write's index mapping entries to be flushed later upon Hudi commit? let's clarify a bit.

Reviewer 3: can you please explain this problem in more detail than 1 sentence.. and even more in comments.. I am not following what the issue is..

This cache can be cleaned once the checkpoint/instant is committed to Hudi.

On job restart or task failover, there is use case that the checkpoint succeeds on Flink while the instant is not committed to Hudi, for DT metadata, the pipeline will recommit

> the instant with the recovered table metadata, while for RLI access, we need to include these special inflight instants on queries, basically, we need to support reading inflight instants on MDT.

Reviewer 2: for RLI, we should also recommit the write status meta just like DT.

> the instant with the recovered table metadata, while for RLI access, we need to include these special inflight instants on queries, basically, we need to support reading inflight instants on MDT.

Reviewer 4: > these special inflight instants

can you define such special inflight instants a bit more clearly? being able to read such inflight instants would be a general capability or a special case for flink MDT use case?

The query will access the in-memory cache first then the MDT RLI index:

![The RLI Access Pattern](./rli-access-pattern.png)

### The Shuffle of RLI Payloads

> In `StreamWrite` operator, the index items are inferred and sent to `IndexWrite` operator in streaming style, the index records are shuffled by `hash(record_key) % num_rli_shards`(the same hashing algorithm of the MDT `RLI index` partitioner),

Reviewer 3: I want to reuse as much common configs as possible. strictly no flink specific table conigs.

   <!--
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->

> # RFC-102: RLI support for Flink streaming

Reviewer 3: is it just RLI or any SI .. Please clarify scope

## Proposers

- @danny0405

## Approvers
 - @geserdugarov
 - @vinothchandar
 - @cshuo

## Status

GH Discussion: https://github.com/apache/hudi/discussions/17452

> Please keep the status updated in `rfc/README.md`.

## Abstract
This RFC aims to introduce RLI support for Flink streaming:

- Impl reliable and performant write and read support for RLI via Flink APIs;
- The RLI impl is engines compatible, for e.g, Flink can access and utilize the RLI written by Spark and vice versa;
- The RLI is global, upserts among partitions is supported; Also support partition level RLI for large fact tables;
- Async compaction for MDT when RLI is enabled; in writer pipeline or table services background job;
- Smart caching of RLI;

> - Clearly set limits for the kind of write throughput supported by RLI (based on certain average response time for the RLI access, like from x0ms to x00ms) via empirical benchmarks;

Reviewer 3: ```suggestion
- Clearly document scale/performance limits for write throughput supported by RLI (based on certain average response time for the RLI access, like from x0ms to x00ms) via empirical benchmarks;
```

> - Ability to be expanded to arbitrary secondary indexing on different columns.

Reviewer 3: Like I have called out many times - we need to cover SI also in one shot. Otherwise, for a Flink table with RLI won't be usable from a SI perspective.

please update RFC title and scope.

## Background
Flink does not support RLI while spark does, this caused inconsistency between engines, for tables migrated from Spark to flink streaming, the index type needs to be switched to either bucket or flink_state , this caused a overhead for users in production.

> Another reason is for multiple partition upserts, currently the only choice is flink_state index, but the flink_state actually costs a lot of memory and can not be shared between different workloads.

Reviewer 3: I would love for us to spell out actual issues like this, such that even a new reader can understand. @danny0405 Please make a pass over the entire RFC like this, before this comment is resolved.

> ## Implementation

Reviewer 3: include diagrams?

The high-level ideas:

- a RLI based index backend will be there to replace the flink_state index;
- a cache of RLI would be introduced to speed the access;

> - a separate index function to write the RLI/SI payloads;

Reviewer 3: ```suggestion
- a separate Flink index function to write the RLI/SI payloads;
```

- the MDT RLI files is written synchronously with the data table data files, the metadata is sent to the coordinator for a final commit to the MDT(after `FILES` partition is ready);
- the MDT compaction is switched to be async and the data files compaction pipeline is reused for less take up of task slots.

> ### The Write

Reviewer 3: ```suggestion
### Implementation
```

### The RLI Access
In `BucketAssigner` operator, the RLI index metadata would be utilized as the index backend, the `BucketAssigner` operator will probe the RLI with the incoming record keys to figure out whether msg is update or insert or delete.
In other words, the RLI index metadata will serve as the same role of the `flink_state` index.

#### The Cache of RLI Access

> We need fast access in streaming to have high throughput(ideally per record access should be < 10ms), thus a general hotspot cache is needed. We will build a in-memory LRU cache by the active upsert records keys, the cache items will be force evictted by a configured memory threshold.

Reviewer 3: Let's specify if we are caching the `key -> location` mappings (record level cache). or parts of the RLI Share (hfile, log file)..  (file level cache).. There are pros/cons to each approach.

We also need a memory cache for the index mappings of current checkpoint because it is not committed to Hudi table yet so invisible.

> This cache can be cleaned once the checkpoint/instant is committed to Hudi(indicates that the index payloads are also committed).

Author: The MDT got its own file based cache, we are gonna try that, here I'm just saying these records are allowed to be evicted instead of invalidate immiediately, we have some basic eviction strategy based on LRU and cache size though.

> On job restart or task failover, there is use case that the checkpoint succeeds on Flink while the instant is not committed to Hudi, for DT metadata, the pipeline will recommit

Reviewer 3: pull into a different section.

We will build an in-memory LRU cache by the active upsert records keys, the cache items will be force evicted by a configured memory threshold.

We also need a memory cache for the index mappings of current checkpoint because it is not committed to Hudi table yet so invisible.
This cache is not allowed to be cleaned until the checkpoint/instant is committed to Hudi(indicates that the index payloads are also committed).

We are caching `key -> location` mappings (record level cache) with LRU style eviction strategy with configurable cache size, the MDT reader also got its native file level cache though.

#### Access of Valid but Uncommitted Instants
On job restart or task failover, there is use case that the checkpoint succeeds on Flink while the instant is not committed to Hudi, for DT metadata, the pipeline will recommit
the instant with the recovered table metadata, because the `BucketAssigner` operator is the upstream operator of `StreamWrite` operator, there is time gap for these inflight instants to recommit,
and we do not want to block the processing of `BucketAssigner`(to wait for the inflight instants to recommit successfully). The suggested solution is

> to include these special inflight instants on RLI access queries, basically, we need to support reading inflight instants on MDT.

Reviewer 3: We should not be adding any "special" new types of instants for this. Once again, please write this out in detail since its not clear what these special instants are, why do they exist when the DT metadata was not written. I have a lot of questions like that.

Author: I add more details about the two-phase commit design of the checkpoint-instant, and such special instants as

>These inflight instants are the ones whose corresponding checkpoint has succeeded, inflight instants without successful checkpoint are not included.

Reviewer 3: are you proposing any changes to the timeline or instant metadata?

Reviewer 5: I also don't understand this clearly. Will sync up w/ @danny0405 and get an understanding.

with incoming record keys to determine the appropriate operation type (insert, update, or delete). In this design, the index serves the same role 
that the `flink_state` index currently serves. Since the existing `BucketAssigner` already supports both **global** and **non-global** index types, 
the global RLI will be used for **global** index configurations, while partitioned RLI will be used for **non-global** configurations.

To optimize index access patterns and avoid caching all index shards in every `BucketAssigner` task, the input records will be shuffled 
by `hash(record_key) % num_index_shards`. This uses the same hashing algorithm as the MDT's index partitioner, ensuring that 
each `BucketAssigner` task only needs to read from a subset of index shards.

#### Index Cache

Streaming workloads require low-latency processing of each record to achieve high throughput. Thus, each record lookup against the index 
should complete really fast. Reading a RLI entry each time for each record will incur 10+ms of latency per record and seriously affect throughput. 

We need two layers of caching to meet this requirement.

> **General-purpose hotspot cache:** The implementation will use an in-memory LRU (Least Recently Used) cache keyed by active upsert record keys. 

Author: We need a cache for records of current checkpoint and for those that are not committed to Hudi, they are not written into files yet.

In general, we read the index cache in streaming style, we are not batching records like spark.

> Cache entries will be evicted when memory usage exceeds a configurable threshold.

Author: We are not checking the log files, we are cleaning the cache with committed checkpoint ids. With OCC concurrent strategy, the cache would anyway been invalited once the conflicts happens after the task got restarted.

**New index mappings cache:** Additionally, a separate memory cache is needed for index mappings created during the current checkpoint. 
These mappings are not yet committed to the Hudi table and are therefore invisible to MDT queries. This cache must not be cleared until the 
corresponding checkpoint/instant is committed to Hudi, which indicates that the index payloads have also been committed. This ensures multiple
records for the same record key (e,g insert to a key, followed by an update within the same commit boundary) are routed consitently to same 
file group, preserving the 1:1 mapping from record key to file group.

The cache stores `key -> location` mappings at the record level, using an LRU eviction strategy with a configurable size limit. 
(Note that the MDT reader also maintains its own native file-level cache.) 

The actual index writes occur in the `IndexWrite` operator and the location from the cache will be propagated downstream from the `BucketAssigner` operator, where cache lookups and MDT queries to determine record locations.
The cache is updated for new records and location changes, while the MDT is queried only for existing key locations.

> The cache update flow is as follows:

Reviewer 3: just confirming, that you are still talking about the mappings cache here..

Author: yes

1. Probe the cache for the key. If found, update the cache entry if the location has changed.
2. If the key is not in the cache, fall back to querying the MDT. If the key exists in the MDT, add it to the cache with its location.
3. If the key does not exist in the MDT either, add the new key and its assigned location to the cache.

#### Job restarts and lost checkpoint acknowledgements

Hudi uses a two-phase commit protocol where each Flink checkpoint corresponds to a Hudi completed instant. During a checkpoint, Flink completes the data writes and collects the Hudi commit metadata. 

> Once the checkpoint acknowledgment event is received, Flink knows the checkpoint completed successfully, and the corresponding Hudi instant can be committed. 

Reviewer 2: Some contexts: 
When a checkpoint is triggered, the snapshot of different operators/coordinators in a flink pipeline are performed concurrently, consider the pipeline: 

`source` -> `calc` -> `sink`

it's possible that the snapshot of `sink` is successful but snapshot of `source` is failed, so during the snapshot, hudi sink only flushes data to fs, sends write status to the coordinator. The committing of table cannot be performed in the snapshot of `sink`, and can only be performed iff the snapshot of the whole pipeline is successful, which ensures the exactly-once semantic. 

When ckp is confirmed as successful, the jobmanager will send checkpoint complete ack to all coordinator/operators. On receiving the ack, hudi writer coordinator will perform the committing of mdt/data table. That's the basic workflow of how flink supports transactional sink. More details see: https://github.com/apache/flink/blob/d30cf4d9ede52b39df5dff39c79b9517228d18f1/flink-core/src/main/java/org/apache/flink/api/common/state/CheckpointListener.java#L24

However, the acknowledgment message is sent asynchronously on a best-effort basis and may be lost in corner cases. A Hudi instant cannot be committed without receiving this acknowledgment.

> During job restarts or task failovers, there are scenarios where a Flink checkpoint succeeds but the corresponding Hudi instant remains uncommitted due to the two-phase commit mechanism:

Author: the flink checkpoint `#snapshot` and its ack is designed to be async, only if a ack msg is received, the checkpoint is ensured to be successful, so we are not commiting to Hudi in `#snapshot` but post the receiving of the ack msg(happens on the JM coordinator).

   7. During the gap between steps 4 and 6, the `BucketAssigner` must access the uncommitted instant because the checkpoint was successful and the data is valid.

For data table (DT) metadata, the pipeline will recommit the instant using the recovered table metadata. However, since the `BucketAssigner` operator is upstream of the `StreamWrite` operator, there is a time gap before these inflight instants can be recommitted. 
We do not want to block `BucketAssigner` processing while waiting for inflight instants to be recommitted. 

The proposed solution is to include these special inflight instants in index access queries—essentially, we need to support reading inflight instants from the MDT. 
Only inflight instants whose corresponding Flink checkpoint has succeeded are included. Inflight instants without a successful checkpoint are excluded. See the Appendix for more details on job/task failover handling.

#### Add Event Time Ordering Value for RLI Payload
For cross-partition updates or deletes, we can not update the RLI directly based on the existing key-location mappings. Currently, the RLI payload only has the key to location mappings without actual ordering value.

> we need to merge the data records to see if the incoming record is a valid update or delete(for valid, it means greater ordering value), that is the behavior for Spark RLI write path, but it is too costly for streaming.

Author: In streaming, we are generally do append write for high perf, reading the existing data set makes the perf similiar with `COW` table, which does not scale well for it's read amplification especially for tables that have frequent cross partiton upserts.

For e.g, for two records `r1:{key: k1, orderingValue: 2, partition: par1}` and `r2:{key: k1, orderingValue: 1, partition: par2}`, `r2` comes behind `r1` in a different commit, 
comparison of just key existence is not enough, we need to also compare the ordering value to see that `r2` is not a valid update and 
not sending the retraction record(delete record) into partition `par1` for payload delete.

> The suggested solution is to store the ordering value into the RLI payload, so that we can compare the ordering value when there is a match of exiting key lookup, to make the decision whether the incoming record is a valid

Author: > Can't we read the RLI shard via FG reader and have that handle all this

The RLI just got a index->location mapping without ordering value. Without the ordring value comparison, we do not know wheter the incoming update needs a location update or just could be dropped.

> - same as the one configured for table. ?

That's the though in my mind as of now.
